core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java

```diff
@@ -67,39 +79,163 @@ public GeoPolygonQueryBuilder addPoint(GeoPoint point) {
         return this;
     }
 
+    public GeoPolygonQueryBuilder points(List<GeoPoint> points) {
+        shell = new ArrayList<>(points);
+        return this;
+    }
+
+    public List<GeoPoint> points() {
+        return shell;
+    }
+
     public GeoPolygonQueryBuilder coerce(boolean coerce) {
+        if (coerce) {
+            this.ignoreMalformed = true;
+        }
         this.coerce = coerce;
         return this;
     }
 
+    public boolean coerce() {
+        return this.coerce;
+    }
+
     public GeoPolygonQueryBuilder ignoreMalformed(boolean ignoreMalformed) {
-        this.ignoreMalformed = ignoreMalformed;
+        if (coerce == false) {
+            this.ignoreMalformed = ignoreMalformed;
+        }
         return this;
     }
 
+    public boolean ignoreMalformed() {
+        return ignoreMalformed;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException errors = null;
+        if (fieldName == null) {
+            errors = QueryValidationException.addValidationError(NAME, "field must not be null", errors);
+        }
+        if (shell.isEmpty()) {
+            errors = QueryValidationException.addValidationError(NAME, "no points defined for geo_polygon query", errors);
+        } else {
+            GeoPoint start = shell.get(0);
+            if (start.equals(shell.get(shell.size() - 1))) {
+                if (shell.size() < 4) {
+                    errors = QueryValidationException.addValidationError(NAME, "too few points defined for geo_polygon query", errors);
+                }
+            } else {
+                if (shell.size() < 3) {
+                    errors = QueryValidationException.addValidationError(NAME, "too few points defined for geo_polygon query", errors);
+                }
+            }
+        }
+        return errors;
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+
+        if (!shell.get(shell.size() - 1).equals(shell.get(0))) {
```

Comment:
I am thinking it is not worth it. Hopefully it won't cause problems to our test infra as we don't check the object after calling toQuery...... should we have dont it in the first place? :)